core/src/main/java/org/kiwix/kiwixmobile/core/main/TableDrawerAdapter.kt

```diff
@@ -0,0 +1,134 @@
+import java.lang.IllegalStateException
+
+class TableDrawerAdapter constructor(private val listener: TableClickListener) :
+  Adapter<ViewHolder>() {
+
+  private var title: String = ""
+  private var selectedPosition: Int = 0
+  private val sections: MutableList<DocumentSection> = mutableListOf()
+
+  fun setSections(sections: List<DocumentSection>) {
+    selectedPosition = 0
+    this.sections.clear()
+    this.sections.addAll(sections)
+  }
+
+  fun setTitle(title: String) {
+    this.title = title
+  }
+
+  override fun getItemViewType(position: Int): Int {
+    return if (position == 0) 0
+    else 1
+  }
+
+  override fun onCreateViewHolder(
+    parent: ViewGroup,
+    viewType: Int
+  ): ViewHolder {
+    val view = parent.inflate(R.layout.section_list, false)
+    if (viewType == 0) return HeaderTableDrawerViewHolder(view)
+    return SectionTableDrawerViewHolder(view)
+  }
+
+  override fun getItemCount(): Int = sections.size + 1
+
+  override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+    when (holder) {
+      is HeaderTableDrawerViewHolder -> {
+        holder.bind(title)
+        holder.itemView.setOnClickListener { v: View ->
+          updateSelection(position)
+          listener.onHeaderClick(v)
+        }
+      }
+      is SectionTableDrawerViewHolder -> {
+        holder.bind(sections[position - 1])
+        holder.itemView.setOnClickListener { v: View? ->
+          updateSelection(position)
+          listener.onSectionClick(v, selectedPosition - 1)
+        }
+      }
+      else -> {
+        throw IllegalStateException("Unknown ViewHolder $holder found")
+      }
+    }
+  }
+
+  private fun updateSelection(position: Int) {
+    if (selectedPosition == position) return
+    val oldPosition = selectedPosition
+    selectedPosition = position
+    notifyItemChanged(selectedPosition)
+    notifyItemChanged(oldPosition)
+  }
+
+  interface TableClickListener {
+    fun onHeaderClick(view: View?)
+    fun onSectionClick(view: View?, position: Int)
+  }
+
+  class HeaderTableDrawerViewHolder(view: View) :
+    BaseViewHolder<String>(view) {
+
+    override fun bind(item: String) {
+      val context = itemView.context
+      titleText.typeface = Typeface.DEFAULT_BOLD
+      if (item.isNotEmpty()) titleText.text = item
+      else {
+        if (context is WebViewProvider) {
+          titleText.text =
+            context.getCurrentWebView()?.title ?: context.getString(R.string.no_section_info)
+        } else titleText.text = context.getString(R.string.no_section_info)
+      }
+    }
+  }
+
+  class SectionTableDrawerViewHolder(view: View) :
+    BaseViewHolder<TableDrawerAdapter.DocumentSection>(view) {
+    override fun bind(
+      item: TableDrawerAdapter.DocumentSection
+    ) {
+      val context = itemView.context
+      val density = context.resources.displayMetrics.density
+      val padding = ((item.level - 1) * TableDrawerAdapter.PADDING_ * density).toInt()
```

Comment:
this is all pretty silly, the original code. Create a dimens value with 16dp and get it here so we don't have to worry about manually multiplying by density, should just be `(item.level-1) * context.resources.getDimension(R.dimen.table_of_contents_padding)`